lib/AST/SubstitutionMap.cpp

```diff
@@ -684,13 +684,23 @@ void SubstitutionMap::profile(llvm::FoldingSetNodeID &id) const {
 }
 
 bool SubstitutionMap::isIdentity() const {
-  for (unsigned i : indices(getReplacementTypes())) {
-    auto replacement = dyn_cast<GenericTypeParamType>(
-      getReplacementTypes()[i]->getCanonicalType(getGenericSignature()));
-    if (!replacement)
-      return false;
-    if (getGenericSignature()->getGenericParamOrdinal(replacement) != i)
-      return false;
-  }
-  return true;
+  GenericSignature *sig = getGenericSignature();
+  unsigned countOfGenericParams = 0;
+  bool hasNonIdentityReplacement = false;
+
+  sig->forEachParam([&](GenericTypeParamType *paramTy, bool isCanonical) {
+    ++countOfGenericParams;
+    if (!isCanonical)
+      return;
+
+    unsigned ordinal = sig->getGenericParamOrdinal(paramTy);
+    auto replacement = getReplacementTypes()[ordinal]->getCanonicalType(sig);
```

Comment:
I think it would be clearer if you just did Type(paramTy).subst(this)

Reply:
That's not the same check, is it? That would also accept SubstitutionMaps that didn't have an entry for that parameter.

Reply:
SubstitutionMaps are always "complete". There's an entry for each parameter